Design/WorkspacePropertyWithIndex.md

```diff
@@ -0,0 +1,95 @@
+# WorkspacePropertyWithIndex Type
+
+This document relates to an algorithm property type which can be used to specify a workspace and desired input index type(s). Index translation will will be handled by the new `IndexInfo` (design [here](https://github.com/mantidproject/documents/blob/master/Design/spectrum_number_and_workspace_index_abstraction.md)) object internally. Badly unit tested code and inconsistencies in validating user indices would be absorbed by this type. This design complements `IndexInfo` and reinforces some of the design concepts. **Note**:The property would then return the workspace and workspace indices (after automatic translation).
+## Motivation
+The `IndexInfo` object recently included in the Mantid framework introduces a strict policy for the definition of and translations between index types. However, this has not yet been reflected at the algorithm property level. There is still user-facing confusion among index types. Algorithms inconsistently capture and use index types without a clear mechanism for defining them e.g spectrum number, spectrum index, workspace index, workspace id, detector index, detector id etc. This is compounded by the soon to be implemented MPI build of Mantid. Indices captured by algorithms must be translated in accordance with data partitioning across MPI ranks. The mechanism in place for this translation exists within *IndexInfo* `IndexInfo::makeIndexSet` but developers must manually make this translation upon obtaining a workspace property.
+
+```cpp
+MatrixWorkspace_sptr inputWs = getProperty("InputWorkspace");
+std::vector<int> indices = getProperty("WorkspaceIndices");
+auto indexSet = inputWs->indexInfo().makeIndexSet(indices);
+```
+ A developer could easily miss this step which would have undefined results in an MPI situation. It would be more sensible to abstract this away from the developer. The `WorkspacePropertyWithIndex` type will obviate the need to perform this translation by doing this internally and returning the correct index set for use by the algorithm. In summary, this property type presents the following benefits:
+
+- Maintaining consistency in how index types are presented and used.
```

Comment:
Maybe add (either separately, or as subitems)
- consistent and unit-tested translation without code duplication
- consistent and unit-tested validation without code duplication
- consistent and simple way of defining index properties without code duplication
- consistent user interface

Edit: I now see that this is mostly discussed on the next section.